### PR TITLE
keep configrations provided by Net::SSH.configuration_for and write s…

### PR DESCRIPTION
**Pull request by sawanoboly**
I want to use ssh options which are loaded from ssh/config.

This PR is replace of #3959

### CODE REVIEW

File: spec/unit/knife/ssh_spec.rb [hunk 1]
```diff
@@ -149,6 +149,71 @@ def self.should_return_specified_attributes
...
+        :keys_only => true
+      )
+    end
+
+    it "should set password" do
+      @knife.config[:ssh_password] = "mysekretpassw0rd"
+      expect(@knife.session_options("the.b.org", nil)).to a_hash_including(:password => "mysekretpassw0rd")
+    end
+
+    it "should placed set identity_file avove password" do
```
Comment (thommay) [inline spec/unit/knife/ssh_spec.rb]: "should prefer identity_file to password"

File: spec/unit/knife/ssh_spec.rb [hunk 2]
```diff
@@ -149,6 +149,71 @@ def self.should_return_specified_attributes
...
+        expect(@knife.session_options("the.b.org", nil)).to a_hash_including(:user => "dog")
+      end
+
+      it "should override port" do
+        @knife.config[:ssh_port] = 9022
+        expect(@knife.session_options("the.b.org", nil)).to a_hash_including(:port => 9022)
+      end
+    end
+
+    context "remove few options from ssh_config provides by NET::SSH" do
```
Comment (thommay) [inline spec/unit/knife/ssh_spec.rb]: "ignore some defaults from Net::SSH"

**Comment by thommay** [comment]
Please fix the failing tests and I've got a couple of grammar nits.

**Comment by sawanoboly** [comment]
Fixed! Thanks @thommay

**Comment by thommay** [comment]
Sweet, :+1:

**Comment by thommay** [comment]
@coderanger did you have concerns about back compat on this?

**Comment by coderanger** [comment]
It is probably fine, just a big enough feature-level change that I think we should all agree this is something we want to support. It means we could end up with a lot more funky usage patterns of `knife ssh` out in the wild. Like what does `Net::SSH` do with `Protocol 1`? Hopefully it is ignored, but if supported by `Net::SSH`, do we want to deal with compat with SSH 1.0? It is an improvement to Principle of Least Surprise UX and I'm :+1: on that basis, but it is opening a can of worms and we should acknowledge that before merge :)

**Comment by sawanoboly** [comment]
If this policy is accepted, I'd like to create another PR to import this policy for `knife bootstrap`.

https://github.com/chef/chef/blob/ececa3438f122d25986fa34557694fa1e86527d2/lib/chef/knife/bootstrap.rb#L388-L400

**Comment by coderanger** [comment]
@sawanoboly bootstrap already uses this code, so no further changes would be needed, ditto it would automatically apply to all the cloud plugins.

**Comment by sawanoboly** [comment]
@coderanger Oh, I see Thanks,

**Comment by sawanoboly** [comment]
@coderanger 

https://github.com/chef/chef/blob/ececa3438f122d25986fa34557694fa1e86527d2/lib/chef/knife/bootstrap.rb#L51

I probably seem that we should remove 'root' by default from bootstrap to use setting from ssh/config?

**Comment by lamont-granquist** [comment]
@coderanger @sawanoboly where we at with this one?

**Comment by lamont-granquist** [comment]
re-pingles?

**Comment by sawanoboly** [comment]
What do you think about reducing configuration from `Net::SSH.configuration_for` ?

- Now, `chef/knife/ssh.rb` picks up options to be used.
- Tthis PR is discarding unused options.

**Comment by thommay** [comment]
I think it's unclear that this PR will ever get merged at this point, so with regrets I'm going to close it. If we want to come back round and come to some kind of consensus please feel free to reopen.
Thanks for your submission and sorry we weren't able to accept it.

**Pull request closed without merging**